### PR TITLE
Colocate Join (#245)

### PR DESCRIPTION
**Pull request by kangkaisen**
[For 245](https://github.com/apache/incubator-doris/issues/245)

Please review, Thanks!

**About Test**
I will add test later.

### CODE REVIEW

File: fe/src/com/baidu/palo/qe/SessionVariable.java [hunk 1]
```diff
@@ -403,6 +407,16 @@ public int getMtDop() {
     public void setMtDop(int mtDop) {
         this.mtDop = mtDop;
     }
+
+    public boolean isDisableColocateJoin()
```
Comment (imay) [inline fe/src/com/baidu/palo/qe/SessionVariable.java]: { should in the same line
public boolean isDisableColocateJoin() {

File: gensrc/thrift/PaloInternalService.thrift [hunk 2]
```diff
@@ -121,6 +121,9 @@ struct TQueryOptions {
 
   // multithreaded degree of intra-node parallelism 
   27: optional i32 mt_dop = 0;
+
+  //colocate join
+  28: optional bool disable_colocate_join = false;
```
Comment (imay) [inline gensrc/thrift/PaloInternalService.thrift]: This flag is useless to BE, no need to add this option

File: fe/src/com/baidu/palo/planner/HashJoinNode.java [hunk 3]
```diff
@@ -114,6 +115,17 @@ public void setDistributionMode(DistributionMode distrMode) {
         this.distrMode = distrMode;
     }
 
+    public boolean isColocate()
+    {
+        return isColocate;
+    }
+
+    public void setColocate(boolean colocate)
+    {
```
Comment (imay) [inline fe/src/com/baidu/palo/planner/HashJoinNode.java]: code style

File: fe/src/com/baidu/palo/planner/DistributedPlanner.java [hunk 4]
```diff
@@ -393,6 +412,83 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
...
+                }
+            }
+            return canColocateJoin(node, leftRoot, rightRoot);
+        }
+
+        return false;
+    }
+
+
+
```
Comment (imay) [inline fe/src/com/baidu/palo/planner/DistributedPlanner.java]: code style: 
should be one empty line

File: fe/src/com/baidu/palo/alter/DecommissionBackendJob.java [hunk 5]
```diff
@@ -433,6 +438,12 @@ public synchronized int tryFinishJob() {
                             continue;
                         }
 
+                        if (olapTable.getColocateTable() != null) {
+                          LOG.info("{} is colocate table, ColocateCloneChecker will handle the tablet clone", olapTable.getName());
```
Comment (imay) [inline fe/src/com/baidu/palo/alter/DecommissionBackendJob.java]: LOG.info will produce too many log? 

change this to debug? or drop this log?

File: fe/src/com/baidu/palo/catalog/Catalog.java [hunk 6]
```diff
@@ -3121,6 +3162,60 @@ private Table createOlapTable(Database db, CreateTableStmt stmt, boolean isResto
...
+                }
+
+                //for colocate child table
+                if (!colocateTable.equalsIgnoreCase(tableName)) {
+                    if (parentTable == null) {
+                      ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_NO_EXIT, colocateTable);
+                    }
+
+                    if (parentTable.type != (TableType.OLAP)) {
+                      ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_OLAP_TABLE);
```
Comment (imay) [inline fe/src/com/baidu/palo/catalog/Catalog.java]: code style: 
4 space indentation

File: fe/src/com/baidu/palo/clone/BackendInfo.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package com.baidu.palo.clone;
+
+public class BackendInfo
```
Comment (imay) [inline fe/src/com/baidu/palo/clone/BackendInfo.java]: code style

**Review by imay — CHANGES_REQUESTED** [review]
some code style case

The code is a bit more, take a closer look at the logic later.

**Comment by kangkaisen** [comment]
Update this commit. 

Because the  change for latest code is big, I will do a full test on this weekend.

**Review by imay — DISMISSED** [review]
You should update your commit log, change it to like "support colocate join" or something, And in the commit log body, you can add more detail about this feature. 

And there are code conflicts to be resolved

**Comment by kangkaisen** [comment]
@imay  I have Updated the commit message and resolved the conflicts. please review. Thanks.

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java [hunk 8]
```diff
@@ -44,9 +45,15 @@ public void analyze(Analyzer analyzer) throws AnalysisException {
             throw new AnalysisException("Properties is not set");
         }
 
-        if (!Catalog.getCurrentCatalog().getAuth().checkGlobalPriv(ConnectContext.get(), PrivPredicate.ALTER)) {
+        if (properties.size() == 1 && properties.containsKey(KEY_COLOCATE_WITH)) {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Currently, alter table colocate_with property need global ALTER Privilege.  so normal user  couldn't alter and we shouldn't grant  normal user  global ALTER Privilege.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 9]
```diff
@@ -2712,6 +2744,13 @@ public void addPartition(Database db, String tableName, AddPartitionClause addPa
             // here we check partition's properties
             singlePartitionDesc.analyze(rangePartitionInfo.getPartitionColumns().size(), null);
 
+            if (olapTable.getColocateTable() != null) {
+                short tableReplicationNum = rangePartitionInfo.idToReplicationNum.entrySet().iterator().next().getValue();
+                if (tableReplicationNum != singlePartitionDesc.getReplicationNum()) {
+                    ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_SAME_REPLICAT_NUM, partitionName);
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: Yes. OK.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 10]
```diff
@@ -3261,6 +3302,59 @@ private Table createOlapTable(Database db, CreateTableStmt stmt, boolean isResto
...
+                        ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_NO_EXIT, colocateTable);
+                    }
+
+                    if (parentTable.type != (TableType.OLAP)) {
+                        ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_OLAP_TABLE);
+                    }
+
+                    HashDistributionInfo parentDistribution = (HashDistributionInfo) ((OlapTable) parentTable).getDefaultDistributionInfo();
+                    if (parentDistribution.getBucketNum() != distributionInfo.getBucketNum()) {
+                        ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_SAME_BUCKNUM);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: OK

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 11]
```diff
@@ -3261,6 +3302,59 @@ private Table createOlapTable(Database db, CreateTableStmt stmt, boolean isResto
...
+                    List<Column> columns = ((HashDistributionInfo) distributionInfo).getDistributionColumns();
+
+                    int columnSize = parentColumns.size();
+                    if (columnSize != columns.size()) {
+                        ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_SAME_DISTRIBUTED_COLUMNS);
+                    }
+
+                    for (int i = 0; i < columnSize; i++) {
+                        if (!parentColumns.get(i).getColumnType().equals(columns.get(i).getColumnType())) {
+                            ErrorReport.reportDdlException(ErrorCode.ERR_COLOCATE_TABLE_MUST_SAME_DISTRIBUTED_COLUMNS);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: OK

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 12]
```diff
@@ -3376,12 +3477,36 @@ private Table createOlapTable(Database db, CreateTableStmt stmt, boolean isResto
...
+                if (olapTable.getColocateTable() != null) {
+                    Long groupId = db.getTable(olapTable.getColocateTable()).getId();
+                    ColocatePersistInfo info = ColocatePersistInfo.CreateForAddTable(tableId, groupId, db.getId());
+                    editLog.logColocateAddTable(info);
+
+                    if (getColocateTableIndex().isColocateParentTable(tableId)) {
+                        List<List<Long>> backendsPerBucketSeq = getColocateTableIndex().getBackendsPerBucketSeq(groupId);
+                        info = ColocatePersistInfo.CreateForBackendsPerBucketSeq(groupId, backendsPerBucketSeq);
+                        editLog.logColocateBackendsPerBucketSeq(info);
+                    }
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: Yes. I see.

I could merge `logColocateAddTable` and `logColocateBackendsPerBucketSeq` to one ,but `createTableWithLock` still have on log. how do we handle this better ?

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 13]
```diff
@@ -4542,6 +4722,34 @@ public void replayRenameTable(TableInfo tableInfo) throws DdlException {
         }
     }
 
+    //the invoker should keep db write lock
+    public void modifyTableColocate(Database db, OlapTable table, String colocateTable) throws DdlException {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: 1 schemaChangeHandler.process has kept  the db write lock, so we needn't add lock here.

2 OK, I will add a check

File: fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 14]
```diff
@@ -395,6 +414,81 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
...
+        }
+
+        if (leftRoot instanceof HashJoinNode && rightRoot instanceof OlapScanNode) {
+            while (leftRoot instanceof HashJoinNode) {
+                if (((HashJoinNode)leftRoot).isColocate()) {
+                    leftRoot = leftRoot.getChild(0);
+                } else {
+                    return false;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java]: OK.

File: fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 15]
```diff
@@ -395,6 +414,81 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
...
+    //the table must be colocate
+    //the colocate group must be stable
+    //the eqJoinConjuncts must contain the distributionColumns
+    private boolean canColocateJoin(HashJoinNode node, PlanNode leftRoot, PlanNode rightRoot) {
+        OlapTable leftTable = ((OlapScanNode) leftRoot).getOlapTable();
+        OlapTable rightTable = ((OlapScanNode) rightRoot).getOlapTable();
+
+        //1 the table must be colocate
+        if (leftTable.getColocateTable() != null &&
+                leftTable.getColocateTable().equalsIgnoreCase(rightTable.getColocateTable())) {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java]: OK

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Why? If one user have ALTER privilege for two tables, it's reasonable that he could make these two tables colocated.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: We must keep metadata keep consistent. If one operation failed, we should guarantee that thing can still work.

Following is a suggest.

If we `createTableWithLock ` first, and log another operation `logChangeTableToColocate`.  And     if the second log failed, we also have complete meta, which have a table without colocate property. 

Can this method work and be implemented easily.

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Yes.  it's reasonable user could later table colocate_with property. but require a global ALTER Privilege is not reasonable. normal users shouldn't have global ALTER Privilege, if they have, they will could read all dbs.  I tested it. So we should check db level Privilege not global Privilege.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: I see. OK

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Yes, you're right, I mistakenly thought that the DB privilege was checked before.

**Comment by imay** [comment]
@kangkaisen 
Could you add some `help doc` to this PR, which help user to use this feature.

**Comment by kangkaisen** [comment]
OK, I will add design doc  and help doc later.

**Comment by kangkaisen** [comment]
Add help doc for colocate join.

**Comment by kangkaisen** [comment]
Hi. @imay Is there any question about this PR?  If not, I will resolve the conflict and rebase the commit.

**Comment by imay** [comment]
> Hi. @imay Is there any question about this PR? If not, I will resolve the conflict and rebase the commit.

Some things have hindered me, so I haven't reviewed yet. I will reply to you earlier next week?

**Comment by kangkaisen** [comment]
OK，Thanks you.

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 16]
```diff
@@ -0,0 +1,573 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * when backend remove, down, and add, balance colocate tablets
+ * some work delegate to {@link CloneChecker}
+ * CloneChecker don't handle colocate tablets
+ */
+public class ColocateTableBalancer extends Daemon {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: OK

File: fe/src/main/java/org/apache/doris/planner/OlapScanNode.java [hunk 17]
```diff
@@ -565,12 +576,20 @@ private void getScanRangeLocations(Analyzer analyzer) throws UserException, Anal
             List<Tablet> tablets = new ArrayList<Tablet>();
             Collection<Long> tabletIds = distributionPrune(selectedTable, partition.getDistributionInfo());
             LOG.debug("distribution prune tablets: {}", tabletIds);
+
+            List<Long> allTabletIds = selectedTable.getTabletIdsInOrder();
             if (tabletIds != null) {
                 for (Long id : tabletIds) {
                     tablets.add(selectedTable.getTablet(id));
+
+                    tabletId2BucketSeq.put(id, allTabletIds.indexOf(id));
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/OlapScanNode.java]: OK

**Comment by kangkaisen** [comment]
Add more comment for `ColocateTableBalancer`

**Comment by kangkaisen** [comment]
Rebase this commit.

**Review by morningman — APPROVED** [review]
LGTM, but I need a further review after merged

**Review by imay — APPROVED** [review]

**Pull request merged**